Fix block issues introduced by the `block.json` updates

m1r0 opened the pull request:
Relates to https://github.com/Automattic/sensei/pull/5782#issuecomment-1271705659

This PR fixes multiple issues that were introduced by the updates to the `block.json` file and the blocks object changes. The most noticeable issue is that the Quiz Block was causing errors when you try to add it, but there are other logic related bugs as well.

The fix reverts to the previous block object structure which has the `block.json` metadata directly in it. It's hard to track down where the code logic uses this metadata, so I've decided to revert it to minimize risk.

### Changes proposed in this Pull Request

* Revert the previous block object structure (before the updates in #5782).

### Testing instructions

* Adding the Quiz Block should work.
* All other blocks should be working as before.

m1r0 committed:
Revert using the metadata object in `block.json`

dadish reviewed (commented):
I'm having a hard time understanding why we are adding a `metadata` property to the blocks' definitions.

```JavaScript
export default {
  ...metadata, // <-- This makes sense. We're merging the block definition from block.json with some additional js properties like edit, save...
  metadata, // <-- This part is confusing. I didn't find a documentation for this property. I don't think it does anything.
}
```

I know the `metadata` property is not being added in this PR, but it was added in the [previous PR](https://github.com/Automattic/sensei/pull/5782) that this PR is addressing. So I figured we could discuss about it here.

***

> The fix reverts to the previous block object structure which has the block.json metadata directly in it. It's hard to track down where the code logic uses this metadata, so I've decided to revert it to minimize risk.

Currently the values from `block.json` are being imported into JS and used to create blocks' settings object and passed to [`registerBlockType`](https://github.com/WordPress/gutenberg/tree/trunk/packages/blocks#registerblocktype) function that we call for each block on the frontend.

But moving forward we need to move to `block.json` files completely and register blocks on the PHP via [`register_block_type`](https://developer.wordpress.org/reference/functions/register_block_type/). This is [the way Gutenberg blocks api is moving](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md). So you work on [previous PR](https://github.com/Automattic/sensei/pull/5782) where you moved all the titles and descriptions of the blocks into `block.json` files is the correct way to move forward.

However as long as we're registering our blocks on the frontend via `registerBlockType` api, we will need to import `block.json` metadata and pass it to `registerBlockType` function. Otherwise the PHP and JS metadata of the blocks might get out of sync.

m1r0 commented:
Hey @dadish, thanks for pointing that out! After further investigation, I think it should be fine removing that object as you suggested. Here's some more context:

The one thing that made me go in this direction was that the Loco Translate plugin not able to translate the strings from `block.json`. Because of that, I wasn't sure if the localization will work at all. ([more info here](https://github.com/Automattic/sensei/pull/5782#issuecomment-1267626522))

To fix this I had to pass down the `block.json` metadata object to `registerBlockType`. If an object is passed to that function instead of the block name, the [`getBlockSettingsFromMetadata`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L280-L284) -> [`translateBlockSettingUsingI18nSchema`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L223-L227) is called and the Loco Translate localization works.

I just checked how things are at Sensei Pro and it looks like the plugin localization works even without passing the metadata object to `registerBlockType`, which is great! Unfortunately, Loco Translate is not able to modify the translated `block.json` strings there as well... I guess that might be a problem with Loco Translate itself.

In the end, I've decided to remove the `metadata` object but I'm not sure how to test the localization of the `block.json` strings. Do you have any idea how to test that? Thanks!

dadish commented:
> To fix this I had to pass down the `block.json` metadata object to `registerBlockType`. If an object is passed to that function instead of the block name, the [`getBlockSettingsFromMetadata`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L280-L284) -> [`translateBlockSettingUsingI18nSchema`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L223-L227) is called and the Loco Translate localization works.

I don't think you need to keep the `block.json` metadata separately in `metadata` property. You can simply pass the whole thing into registerBlockType function.

```JavaScript
// block's index.js file
export default {
    ...metadata,
    edit: function () {},
    save: function () {}
}

// registering the block
blocks.forEach( ( block ) => {
  registerBlockType( block, block );
});
```
This should  serve the same function.

> In the end, I've decided to remove the `metadata` object but I'm not sure how to test the localization of the `block.json` strings. Do you have any idea how to test that? Thanks!

Unfortunately not. I did some local testing and tried different things, but could not figure out how to test language strings in `block.json` files.

m1r0 commented:
> I don't think you need to keep the `block.json` metadata separately in `metadata` property. You can simply pass the whole thing into registerBlockType function.
>```js
>blocks.forEach( ( block ) => {
>  registerBlockType( block, block );
>});
>```

I did try this option but the localization is still not working.

I think this is because of how `registerBlockType` works - the first `block` argument ([`blockNameOrMetadata`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L259)) gets localized, but the second `block` argument ([`settings`](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L259)) is not localized by `translateBlockSettingUsingI18nSchema` and it [overwrites](https://github.com/WordPress/gutenberg/blob/154a8b876a754e0ee678549687afe3d49baa58d7/packages/blocks/src/api/registration.js#L297-L298) the first because it has the same keys... 🙂 I guess we should follow what `registerBlockType` is expecting to receive.

I can think of 3 options to resolve this:

1. Have a separate `metadata` key that only holds the `block.json` data and pass that as the first param to `registerBlockType`. This makes Loco Translate happy for some reason and the localization is testable.
2. Use the same approach as Sensei Pro which is to pass the block name as the first param to `registerBlockType` and the `block.json` data as the second param. The localization still works (not sure how) when you install the released package, but I'm not able to test locally using Loco Translate.
3. Refactor how the blocks are registered.

Is it safe to assume that if we use the same approach as Sensei Pro the localization will work? Worst case, the blocks title/description/keywords won't be localized and we will have to revisit.

We could also have the redundant `metadata` key for now and figure out how to refactor later. What do you think?

dadish commented:
> Is it safe to assume that if we use the same approach as Sensei Pro the localization will work? Worst case, the blocks title/description/keywords won't be localized and we will have to revisit.

I don't know if the translates will work the same way in Sensei as in Sensei Pro. But I agree that it's better if we have the similar ways of registering blocks and handling i18n for blocks.

> We could also have the redundant metadata key for now and figure out how to refactor later. What do you think?

Agree, let's keep it. My only concern is that it's not clear from the code why the redundant `metadata` is needed. A person like me might think it's not needed and remove it. If we decide to keep it, could you at least add a comment explaining it a bit. You can add a comment only in `assets/blocks/register-sensei-blocks.js` file. That should be enough.

m1r0 committed:
Move the featured video block title and description to `block.json`

m1r0 committed:
Fix `block.json` localization

m1r0 commented:
> Agree, let's keep it. My only concern is that it's not clear from the code why the redundant metadata is needed. A person like me might think it's not needed and remove it. If we decide to keep it, could you at least add a comment explaining it a bit. You can add a comment only in `assets/blocks/register-sensei-blocks.js` file. That should be enough.

Yep, makes total sense! I've added a few comments but also had to add a [workaround for the settings object to not overwrite the localized metadata](https://github.com/Automattic/sensei/pull/5905/commits/778eae4da570b016216385ce81035206b9349d48#diff-d0270f28f071080f818a9e240ae55fae85d5cf2384eaf8d9ba14ee88c2be0f55R30-R34).

I think this whole approach of passing metadata and settings should be refactored, so the settings are not holding metadata as well. Unfortunately this might require a considerable amount of time. I've added a [`@todo`](https://github.com/Automattic/sensei/pull/5905/commits/778eae4da570b016216385ce81035206b9349d48#diff-d0270f28f071080f818a9e240ae55fae85d5cf2384eaf8d9ba14ee88c2be0f55R19) for now because it's out of scope.

m1r0 commented:
Thanks @dadish for all the help!

yscik commented:
5631824-zd-woothemes 
5636709-zd-woothemes
